lib/puppet/configurer.rb

```diff
@@ -225,12 +225,22 @@ def run(options = {})
         # mode. We shouldn't try to do any failover in that case.
         if options[:catalog].nil? && do_failover
           server, port = find_functional_server
-          if server.nil?
-            raise Puppet::Error, _("Could not select a functional puppet master from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
-          else
-            #TRANSLATORS 'server_list' is the name of a setting and should not be translated
-            Puppet.debug _("Selected puppet server from the `server_list` setting: %{server}:%{port}") % { server: server, port: port }
-            report.master_used = "#{server}:#{port}"
+          begin
+            if server.nil?
+              raise Puppet::Error, _("Could not select a functional puppet server from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
+            else
+              #TRANSLATORS 'server_list' is the name of a setting and should not be translated
+              Puppet.debug _("Selected puppet server from the `server_list` setting: %{server}:%{port}") % { server: server, port: port }
+              report.master_used = "#{server}:#{port}"
+            end
+          rescue Puppet::Error => detail
+            if Puppet[:usecacheonfailure]
+              options[:pluginsync] = false
+              @running_failure = true
+              Puppet.log_exception(detail)
+            else
+              raise detail
+            end
           end
           Puppet.override(server: server, serverport: port) do
```

Comment:
One nit, if `server` and `port` are nil then we push nil values onto the context, which reintroduces the issue in https://tickets.puppetlabs.com/browse/PUP-9076 (trying to connect to port 0 for the node and report request)

```
$ bx puppet agent --no-daemonize --verbose --onetime  --server_list www.example.com,hatted-absentia.delivery.puppetlabs.net:888 --http_debug --http_connect_timeout 1s
opening connection to www.example.com:8140...
opening connection to hatted-absentia.delivery.puppetlabs.net:888...
Error: Could not select a functional puppet server from server_list: 'www.example.com,hatted-absentia.delivery.puppetlabs.net:888'
opening connection to :0...
Warning: Unable to fetch my node definition, but the agent run will continue:
Warning: Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
Info: Using cached catalog from environment 'production'
Info: Caching catalog for localhost
Info: Applying configuration version '1602269179'
Notice: Applied catalog in 0.01 seconds
opening connection to :0...
Error: Could not send report: Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
```

So I'm thinking if `server.nil?` then we should set `server` to the first host entry in `Puppet[:server_list]` and set `port` either the port entry or default to `Puppet[:serverport]`. Something like:

```ruby
if server.nil?
  server =   Puppet[:server_list].first[0]
  port  = Puppet[:server_list].first[1] || Puppet[:masterport]
  raise ...
```

Reply:
After the change the log looks like bellow:
```
[root@movable-vioiln ~]# puppet agent --no-daemonize --verbose --onetime  --server_list www.example.com:1479,hatted-absentia.delivery.puppetlabs.net:888 --http_debug>
opening connection to www.example.com:1479...
opening connection to hatted-absentia.delivery.puppetlabs.net:888...
Error: Could not select a functional puppet server from server_list: 'www.example.com:1479,hatted-absentia.delivery.puppetlabs.net:888'
opening connection to www.example.com:1479...
Warning: Unable to fetch my node definition, but the agent run will continue:
Warning: execution expired
Info: Using cached catalog from environment 'production'
Info: Caching catalog for movable-vioiln.delivery.puppetlabs.net
Info: Applying configuration version '1602488455'
Notice: Applied catalog in 0.01 seconds
opening connection to www.example.com:1479...
Error: Could not send report: execution expired
```